web/src/app/components/bar-chart/bar-chart.component.ts

```diff
@@ -0,0 +1,341 @@
+   * pushes for dataAll and of complete pushes for dataComplete. Both arrays
+   * are in ascending order by startTime.
+   * @param pushInfos Array for one push def
+   */
+  private populateData(pushInfos: step189_2020.IPushInfo[]): void {
+    if (!pushInfos) { return; }
+    pushInfos.reverse().forEach(pushInfo => {
+      if (!pushInfo) { return; }
+      const states = pushInfo.stateInfo;
+      if (!states) { return; }
+      const pushID = pushInfo.pushHandle;
+      if (!pushID) { return; }
+      const endState = states[states.length - 1].state;
+      if (!endState) { return; }
+      const pushStartTime = states[0].startTimeNsec;
+      if (!pushStartTime) { return; }
+      const pushEndTime = states[states.length - 1].startTimeNsec;
+      if (!pushEndTime) { return; }
+
+      // Filter pushes with only one state (0 duration) and endState which 
+      // should not be considered.
+      if (states.length > 1) { 
+        if (endState !== BarChartComponent.ABANDONED_STATE_TAG 
+          && endState !== BarChartComponent.BUILT_STATE_TAG
+          && endState !== BarChartComponent.BUILT_UNCHANGED_TAG
+          && endState !== BarChartComponent.SET_DEFAULT_VERSION_STATE_TAG) {
+          const startTime = formatDate((+pushStartTime / BarChartComponent.NANO_TO_MILLI), 
+            BarChartComponent.DATE_FORMAT, BarChartComponent.DATE_LOCALE);
+          const duration = (+pushEndTime - +pushStartTime) / 
+            BarChartComponent.NANO_TO_SECS / BarChartComponent.SECS_TO_HRS;
+          const thePush: BarChartData = {
+            pushID,
+            state: endState,
+            startTime,
+            duration
+          };
+          this.totalDuration.push(duration);
+          this.dataAll.push(thePush);
+          
+          if (endState === BarChartComponent.COMPLETE_STATE_TAG) {
+            this.dataComplete.push(thePush);
+            this.completeDuration.push(duration);
+          }
+        }
+      }
+    })
+  }
+
+  /**
+   * This function scales the x-axis and y-axis and initializes empty focus 
+   * and brush elements.
+   * 
+   * Structure of the SVG:
+   * focus:
+   *   - yAxisFocus, xAxisFocus
+   *   - title of y-axis
+   * brush:
+   *   - xAxisBrush
+   */
+  private initialChart(): void {
+    const element = this.barChartContainer.nativeElement;
+    const elementWidth = element.clientWidth;
+    const elementHeight = element.clientHeight;
+    const marginFocus = { top: 60, right: 90, bottom: 150, left: 70 };
+    const marginBrush = { top: 370, right: 60, bottom: 20, left: 60};
+
+    this.heightFocus = elementHeight;
+    this.heightBrush = 30;
+    this.width = elementWidth;
+
+    this.svg = (d3
+      .select(element)
+      .append('svg') as d3SVG)
+      .attr('width', elementWidth)
+      .attr('height', elementHeight);
+
+    this.focus = this.svg
+      .append('g')
+      .attr('transform','translate(0, 0)');
+    this.brush = this.svg
```

Comment:
It's not clear why this is called brush.

Reply:
"Brushing is the interactive specification a one- or two-dimensional selected region using a pointing gesture, such as by clicking and dragging the mouse." [1] Or maybe I should called it `context`?

[1] https://github.com/d3/d3-brush

Reply:
I think we can keep the name of brush if we clearly explain what is referring to.